include/rmm/mr/device/arena_memory_resource.hpp

```diff
@@ -0,0 +1,287 @@
+   *
+   * @param bytes The size in bytes of the allocation
+   * @param stream The stream to associate this allocation with
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override
+  {
+    if (bytes <= 0) return nullptr;
+
+    bytes = rmm::detail::align_up(bytes, allocation_alignment);
+    RMM_EXPECTS(
+      bytes <= get_maximum_allocation_size(), rmm::bad_alloc, "Maximum allocation size exceeded");
+
+    if (arena::handles_size(bytes)) {
+      return get_arena(stream).allocate(bytes);
+    } else {
+      RMM_ASSERT_CUDA_SUCCESS(cudaStreamSynchronize(stream));
+      return upstream_mr_->allocate(bytes, cudaStreamLegacy);
+    }
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by `p`.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @param stream Stream on which to perform deallocation
+   */
+  void do_deallocate(void* p, std::size_t bytes, cudaStream_t stream) override
+  {
+    if (p == nullptr || bytes <= 0) return;
+
+    bytes = rmm::detail::align_up(bytes, allocation_alignment);
+    if (arena::handles_size(bytes)) {
+      if (!get_arena(stream).deallocate(p, bytes, stream)) {
+        deallocate_across_arenas(p, bytes, stream);
+      }
+    } else {
+      RMM_ASSERT_CUDA_SUCCESS(cudaStreamSynchronize(stream));
+      upstream_mr_->deallocate(p, bytes, cudaStreamLegacy);
+    }
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by `p` that was allocated in a different arena.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @param stream Stream on which to perform deallocation
+   */
+  void deallocate_across_arenas(void* p, std::size_t bytes, cudaStream_t stream)
+  {
+    RMM_ASSERT_CUDA_SUCCESS(cudaStreamSynchronize(stream));
+
+    read_lock lock(mtx_);
+
+    if (is_default_stream(stream)) {
+      auto id = std::this_thread::get_id();
+      for (auto& kv : thread_arenas_) {
+        if (kv.first != id && kv.second.deallocate(p, bytes)) return;
+      }
+    } else {
+      for (auto& kv : stream_arenas_) {
+        if (kv.first != stream && kv.second.deallocate(p, bytes)) return;
+      }
+    }
+
+    // Allocation not found.
+    RMM_LOGGING_ASSERT(0);
+  }
+
+  /**
+   * @brief Get the maximum size of allocations supported by this memory resource.
+   *
+   * Note this does not depend on the memory size of the device. It simply returns the maximum
+   * value of `std::size_t`.
+   *
+   * @return std::size_t The maximum size of a single allocation supported by this memory resource
+   */
+  std::size_t get_maximum_allocation_size() const
```

Comment:
Note I only added this function in classes that derive from `stream_ordered_memory_resource`. I don't think you need it here.

Reply:
Removed.